### DISCUSSION TITLE
Fire TV media player support

### DISCUSSION THREAD

**Discussion by genman** [schwabe/ics-openvpn#1418] [Ideas]
The OpenVPN app works great on the Fire TV but the UI requires an addition of a mouse pointer app, and a physical mouse–bluetooth, to navigate and use.

There's several tutorials about downloading mouse support, so it's a common enough end-user issue.

It seems it would be easy enough to support the remote command input. Unfortunately the company I work with is fairly restrictive in contributing code, so I can't be helping directly.

My guess is adding keyboard support is probably "good enough" for most users. By that I mean the remote works as a keyboard to change focus over the buttons. The remote does not pass the tab character so the focus needs to be manually managed by the app.

Here's an example of the keyboard events that are passed:
https://www.developer.amazon.com/docs/fire-tv/remote-input.html

My guess is the OnClickListener would need to add this interface and implementation:
https://developer.android.com/reference/android/view/View.OnKeyListener

It might be wise to monitor for "controller actions" instead of generic events.
https://developer.android.com/training/game-controllers/controller-input

**Comment by schwabe** [comment]
The Android TV UI support in general is not very good. The problem is that proper support requires an almost unique UI just for TVs. That is a lot of work and this is a spare time project, so my motivation to spend that time for that kind of UI design is pretty much non-existant. 

Patches are always welcome but just to be honest, you will probably not see much from my side since there are other more interesting tasks in my app that I still want to work on instead.

**Reply by genman** [reply]
Do we really need to create a whole new UI?

I understand this is a part time deal and creating a new UI seems like a lot of work. Is there some shortcut/key work-arounds like I suggest that can at least handle various connect workflows?

On my side there's lawyers and managers I would need to convince to even get to work on this.

**Reply by schwabe** [reply]
Well the current UI already supports as good keyboard/controller navigation as the framework library gives me. And that is not really that well. Adding better control with the current UI would be as complicated as just adding some addiotinal UI for Android TV.

**Reply by genman** [reply]
> Well the current UI already supports as good keyboard/controller navigation as the framework library gives me. And that is not really that well. Adding better control with the current UI would be as complicated as just adding some addiotinal UI for Android TV.

It could just be something like, if the user pushes up on the remote, you connect. If they push down it disconnects. Left it exits the app, right it shows the connection status. It doesn't need to cover every aspect of the existing UI.

What's not as necessary is adding and removing VPN configurations as it's done infrequently enough.

**Reply by jvirg** [reply]
An app for firestick that would be cool but very unlikely

**Reply by schwabe** [reply]
Yeah I just have limited time and when I have free spare time, there is only so much that I can work on. And I don't want to promise something that I will not add because it has low priority. I just want to be honest about it.
